clone.go

```diff
@@ -0,0 +1,262 @@
+https://en.wikipedia.org/wiki/Golden_Gate_Cloning
+https://www.neb.com/applications/cloning-and-synthetic-biology/dna-assembly-and-cloning/golden-gate-assembly
+
+GoldenGate is a particular kind of restriction enzyme cloning reaction that you can do
+in a single tube and that is extraordinarily efficient (up to 50 parts) and is popular
+for new modular DNA part toolkits. Users can easily simulate GoldenGate assembly reactions
+with just their input fragments + the enzyme name.
+
+Let's build some DNA!
+
+Keoni
+
+******************************************************************************/
+
+// CloneSequence is a simple struct that can carry a circular or linear DNA sequence.
+type CloneSequence struct {
+	Sequence string
+	Circular bool
+}
+
+// Overhang is a struct that represents the ends of a linearized sequence where Enzymes had cut.
+type Overhang struct {
+	Length   int
+	Position int
+	Forward  bool
+}
+
+// Fragment is a struct that represents linear DNA sequences with sticky ends.
+type Fragment struct {
+	Sequence        string
+	ForwardOverhang string
+	ReverseOverhang string
+}
+
+// Enzyme is a struct that represents restriction enzymes.
+type Enzyme struct {
+	EnzymeName        string
+	RegexpFor         *regexp.Regexp
+	RegexpRev         *regexp.Regexp
+	EnzymeSkip        int
+	EnzymeOverhangLen int
+	Directional       bool
+}
+
+/******************************************************************************
+
+Base cloning functions begin here.
+
+******************************************************************************/
+
+// https://qvault.io/2019/10/21/golang-constant-maps-slices/
+func getBaseRestrictionEnzymes() map[string]Enzyme {
+	// Eventually, we want to get the data for this map from ftp://ftp.neb.com/pub/rebase
+	enzymeMap := make(map[string]Enzyme)
+
+	// Build default enzymes
+	enzymeMap["BsaI"] = Enzyme{"BsaI", regexp.MustCompile("GGTCTC"), regexp.MustCompile("GAGACC"), 1, 4, true}
+	enzymeMap["BbsI"] = Enzyme{"BbsI", regexp.MustCompile("GAAGAC"), regexp.MustCompile("GTCTTC"), 2, 4, true}
+	enzymeMap["BtgZI"] = Enzyme{"BtgZI", regexp.MustCompile("GCGATG"), regexp.MustCompile("CATCGC"), 10, 4, true}
+	enzymeMap["BsmBI"] = Enzyme{"BsmBI", regexp.MustCompile("CGTCTC"), regexp.MustCompile("GAGACG"), 1, 4, true}
+
+	// Return EnzymeMap
+	return enzymeMap
+}
+
+// RestrictionEnzymeCut cuts a given sequence with an enzyme represented by the enzyme's name.
+func RestrictionEnzymeCut(seq CloneSequence, enzymeStr string) ([]Fragment, error) {
+	enzymeMap := getBaseRestrictionEnzymes()
+	if _, ok := enzymeMap[enzymeStr]; ok == false {
+		return []Fragment{}, errors.New("Enzyme " + enzymeStr + " not found in enzymeMap")
+	}
+	enzyme, _ := enzymeMap[enzymeStr]
+	return RestrictionEnzymeCutEnzymeStruct(seq, enzyme), nil
+}
+
+// RestrictionEnzymeCutEnzymeStruct cuts a given sequence with an enzyme represented by an Enzyme struct.
+func RestrictionEnzymeCutEnzymeStruct(seq CloneSequence, enzyme Enzyme) []Fragment {
```

Comment:
```suggestion
func CutSequenceWithEnzyme(seq CloneSequence, enzyme Enzyme) []Fragment {
```

Reply:
Also unit test ples.

The functions in this file are each so complex that a break would be close to impossible to debug without a test for each function.

Reply:
Gotcha, more tests. Renamed.